envs/monkey_zoo/blackbox/tests/basic_test.py

```diff
@@ -14,11 +13,12 @@
 
 class BasicTest(object):
 
-    def __init__(self, name, island_client, config_parser, analyzers, timeout, log_handler):
+    def __init__(self, name, island_client, config_parser, analyzers, timeout, post_exec_analyzers, log_handler):
         self.name = name
         self.island_client = island_client
         self.config_parser = config_parser
         self.analyzers = analyzers
+        self.post_exec_analyzers = post_exec_analyzers
```

Comment:
BasicTest class is already too convoluted as it is. We don't need `post_exec_analyzers`, `pre_exec_analyzers` or anything similar here. Please create `PerformanceTest(BasicTest)` class and override `run` method to use analyzers after all monkeys are ran and are dead. Also, that class should contain all this test specific consts, use monkey island requests to send timed requests and aggregate all times. Those times should then be sent to the analyzer to get bool result.

Reply:
I disagree. 
The reason we chose to have Analyzers as a generic concept was to be able to easily add analyzers. The first time we needed to add one, the old design wasn't good enough. 

I think that WHEN the analyzer runs is a really important distinction of an analyzer;
* An Analyzer can either run during the test until some condition is met (like the report showing something specific or all monkeys reached the target machines like the CommunicationAnalyzer) - This category is called `Analyzer` today, but we can define this more specifically to be `PolledAnalyzer` or `InProgressAnalyzer` or `DuringAnalyzer` etc.
* Or an Analyzer can run after all Monkeys have died - This category is called `PostExecAnalyzer` but we can find a better name for it as well

Think about 2-3 analyzers that we might add and you'll see that they fit nicely into these categories.

What we can do to express this in design is: 

1. Have the Analyzer abc define an abstract property called "isPolled"
2. Have 2 sub-classes abc's of Analyzer - one for `during` and one for `post`. You pass all the analyzers into the analyzers list and then the BasicTest class chooses which analyzers to run when based on their type
3. Leave it as it is

I don't think we should express this in the `*Test` object design since it seems much less elegant.

Reply:
My view is that communication test has a different workflow than performance test. Performance test shouldn't particulary care if monkeys communicated back or not, it should just wait until all monkeys are dead and then run the analyzers. 
So one test doesn't use post execution analyzers, but they must pass a parameter and another test does not use runtime analyzers, but user still has to pass this as a parameter. 
Ultimately an analyzer analyzes island state at the time it is ran and I'd like to keep it that simple. It should be for the test to decide when to run which analyzers. 
What if we have a test where we run monkey, then analyze, then change config and then run monkey again and finally analyze one more time? We'll have to add `first_run_runtime_analyzers`, `first_run_post_run_analyzers`, `second_run_runtime_analyzers`, `second_run_post_run_analyzers`?
TL;DR I'd rather have simple analyzers and different tests than having a single bloated test class to fit all possible scenarios.
Since having one useless parameter isn't that bad I'd prefer 3, but we should be wary of making "swiss army knife" classes.

Reply:
That's not really the case since the performance test DOES care if Monkeys communicated back.

I agree that the BasicTest class is getting a little bloated - I think we should pin this and the next time we approach it we should have a re-think about the design. I feel like it mostly does Graph Creation which is fine